// @liskhq/lisk-transactions/dist-node/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseTransactionSchema = exports.validateTransaction = exports.signMultiSignatureTransactionWithPrivateKey = exports.signTransactionWithPrivateKey = exports.signMultiSignatureTransaction = exports.signTransaction = exports.getSigningBytes = exports.getBytes = exports.convertLSKToBeddows = exports.convertBeddowsToLSK = exports.DEFAULT_SIGNATURE_BYTE_SIZE = exports.DEFAULT_NUMBER_OF_SIGNATURES = exports.DEFAULT_MIN_FEE_PER_BYTE = exports.computeMinFee = void 0;
var fee_1 = require("./fee");
Object.defineProperty(exports, "computeMinFee", { enumerable: true, get: function () { return fee_1.computeMinFee; } });
Object.defineProperty(exports, "DEFAULT_MIN_FEE_PER_BYTE", { enumerable: true, get: function () { return fee_1.DEFAULT_MIN_FEE_PER_BYTE; } });
Object.defineProperty(exports, "DEFAULT_NUMBER_OF_SIGNATURES", { enumerable: true, get: function () { return fee_1.DEFAULT_NUMBER_OF_SIGNATURES; } });
Object.defineProperty(exports, "DEFAULT_SIGNATURE_BYTE_SIZE", { enumerable: true, get: function () { return fee_1.DEFAULT_SIGNATURE_BYTE_SIZE; } });
var format_1 = require("./format");
Object.defineProperty(exports, "convertBeddowsToLSK", { enumerable: true, get: function () { return format_1.convertBeddowsToLSK; } });
Object.defineProperty(exports, "convertLSKToBeddows", { enumerable: true, get: function () { return format_1.convertLSKToBeddows; } });
var sign_1 = require("./sign");
Object.defineProperty(exports, "getBytes", { enumerable: true, get: function () { return sign_1.getBytes; } });
Object.defineProperty(exports, "getSigningBytes", { enumerable: true, get: function () { return sign_1.getSigningBytes; } });
Object.defineProperty(exports, "signTransaction", { enumerable: true, get: function () { return sign_1.signTransaction; } });
Object.defineProperty(exports, "signMultiSignatureTransaction", { enumerable: true, get: function () { return sign_1.signMultiSignatureTransaction; } });
Object.defineProperty(exports, "signTransactionWithPrivateKey", { enumerable: true, get: function () { return sign_1.signTransactionWithPrivateKey; } });
Object.defineProperty(exports, "signMultiSignatureTransactionWithPrivateKey", { enumerable: true, get: function () { return sign_1.signMultiSignatureTransactionWithPrivateKey; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateTransaction", { enumerable: true, get: function () { return validate_1.validateTransaction; } });
__exportStar(require("./constants"), exports);
var schema_1 = require("./schema");
Object.defineProperty(exports, "baseTransactionSchema", { enumerable: true, get: function () { return schema_1.baseTransactionSchema; } });
