// @liskhq/lisk-tree/dist-node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MerkleTree = exports.regularMerkleTree = void 0;
const verify_proof_1 = require("./merkle_tree/verify_proof");
const calculate_1 = require("./merkle_tree/calculate");
const right_witness_1 = require("./merkle_tree/right_witness");
const merkle_tree_1 = require("./merkle_tree/merkle_tree");
Object.defineProperty(exports, "MerkleTree", { enumerable: true, get: function () { return merkle_tree_1.MerkleTree; } });
const utils_1 = require("./merkle_tree/utils");
exports.regularMerkleTree = {
    verifyProof: verify_proof_1.verifyProof,
    verifyDataBlock: verify_proof_1.verifyDataBlock,
    calculateRootFromUpdateData: calculate_1.calculateRootFromUpdateData,
    calculateMerkleRoot: utils_1.calculateMerkleRoot,
    calculateMerkleRootWithLeaves: utils_1.calculateMerkleRootWithLeaves,
    MerkleTree: merkle_tree_1.MerkleTree,
    calculateRootFromRightWitness: right_witness_1.calculateRootFromRightWitness,
    verifyRightWitness: right_witness_1.verifyRightWitness,
    calculateRightWitness: right_witness_1.calculateRightWitness,
};
